pkg/upstream/cluster/loadbalancer.go

```diff
@@ -333,3 +338,184 @@ func hostWeightsAreEqual(hosts []types.Host) bool {
+// ** node(0)
+// *** leaf-0(0)
+// *** leaf-1(1)
+// ** node(2)
+// *** leaf-2(2)
+// *** leaf-3(3)
+// @endmindmap
+// ```
+//
+// As shown, healthy host index are in quotes, reprecenting data saving in segment tree.
+// When host are not healthy(for example index 2), call tree's Update function and the tree will be like:
+//
+// ```plantuml
+// @startmindmap
+// * root-node(0)
+// ** node(0)
+// *** leaf-0(0)
+// *** leaf-1(1)
+// ** node(3)--node data is changed, always save healthy host index
+// *** leaf-2(2)--unhealthy
+// *** leaf-3(3)
+// @endmindmap
+// ```
+func newMaglevLoadBalancer(info types.ClusterInfo, set types.HostSet) types.LoadBalancer {
+	var table *maglev.Table
+	var tree *segmenttree.Tree
+	names := []string{}
+	for _, host := range set.Hosts() {
+		names = append(names, host.AddressString())
+	}
+	if len(names) != 0 {
+		table = maglev.New(names, maglev.SmallM)
+		// build tree, devide hash
+		nodes := []segmenttree.Node{}
+		step := math.MaxUint64 / uint64(len(names))
+		for index := range names {
+			nodes = append(nodes, segmenttree.Node{
+				Value:      index,
+				RangeStart: uint64(index) * step,
+				RangeEnd:   uint64(index+1) * step,
+			})
+		}
+		updateFunc := getSegmentTreeUpdateFunc(set)
+		tree = segmenttree.NewTree(nodes, updateFunc)
+	}
+
+	mgv := &maglevLoadBalancer{
+		hosts:           set,
+		maglev:          table,
+		fallbackSegTree: tree,
+	}
+
+	return mgv
+}
+
+// getSegmentTreeUpdateFunc return SegmentTreeUpdateFunc with loadbalancer logic.
+// Though this function, segment tree's non-leaf node's data save healthy host index of left & right sub-tree.
+//
+// When both left & right sub-tree are healthy, current node data is left sub-tree's data, for this function will
+// return right away when found sub-tree is healthy.
+// If both left and right sub-tree are unhealthy, current node data is nil.
+func getSegmentTreeUpdateFunc(set types.HostSet) segmenttree.SegmentTreeUpdateFunc {
+	return func(leftChildNodeValue, rightChildNodeValue interface{}) interface{} {
+		if leftChildNodeValue != nil {
+			leftIndex, ok := leftChildNodeValue.(int)
+			if !ok {
+				return nil
+			}
+			if set.Hosts()[leftIndex].Health() {
+				return leftIndex
+			}
+		}
+
+		if rightChildNodeValue != nil {
+			rightIndex, ok := rightChildNodeValue.(int)
+			if !ok {
+				return nil
+			}
+
+			if set.Hosts()[rightIndex].Health() {
+				return rightIndex
+			}
+		}
+
+		return nil
+	}
+}
+
+type maglevLoadBalancer struct {
+	hosts           types.HostSet
+	maglev          *maglev.Table
+	fallbackSegTree *segmenttree.Tree
+}
+
+func (lb *maglevLoadBalancer) ChooseHost(context types.LoadBalancerContext) types.Host {
+	// host empty, maglev info may be nil
+	if lb.maglev == nil {
+		return nil
+	}
+
+	route, ok := mosnctx.Get(context.DownstreamContext(), types.ContextKeyDownStreamRouter).(api.Route)
+	if !ok {
+		return nil
+	}
+
+	hash := route.RouteRule().Policy().HashPolicy().GenerateHash(context.DownstreamContext())
```

Comment:
the value of `Policy().HashPolicy()` maybe is nil.

Reply:
fixed